acpi/src/lib.rs

```diff
@@ -95,189 +100,173 @@ pub enum AcpiError {
-
-                /*
-                 * We've already validated the FADT to get the DSDT out, so it doesn't need to be done again.
-                 */
-                self.sdts
-                    .insert(Signature::FADT, Sdt { physical_address, length: header.length, validated: true });
-            }
-            Signature::SSDT => {
-                self.ssdts.push(AmlTable::new(physical_address, header.length));
+    /// Finds and returns the DSDT AML table, if it exists.
+    pub fn get_dsdt(&self) -> AcpiResult<AmlTable> {
+        self.find_table::<fadt::Fadt>().and_then(|fadt| {
+            struct Dsdt;
+            impl AcpiTable for Dsdt {
+                const SIGNATURE: Signature = Signature::DSDT;
+
+                fn header(&self) -> &sdt::SdtHeader {
+                    // ### Safety: DSDT will always be valid for an SdtHeader at its `self` pointer.
+                    unsafe { &*(self as *const Self as *const sdt::SdtHeader) }
+                }
             }
-            signature => {
-                self.sdts.insert(signature, Sdt { physical_address, length: header.length, validated: false });
-            }
-        }
 
-        Ok(())
+            let dsdt_address = fadt.dsdt_address()?;
+            let dsdt = unsafe { read_table::<H, Dsdt>(self.handler.clone(), dsdt_address)? };
+
+            Ok(AmlTable::new(dsdt_address, dsdt.header().length))
+        })
     }
 
-    /// Create a mapping to a SDT, given its signature. This validates the SDT if it has not already been
-    /// validated.
-    ///
-    /// ### Safety
-    /// The table's memory is naively interpreted as a `T`, and so you must be careful in providing a type that
-    /// correctly represents the table's structure. Regardless of the provided type's size, the region mapped will
-    /// be the size specified in the SDT's header. Providing a `T` that is larger than this, *may* lead to
-    /// page-faults, aliasing references, or derefencing uninitialized memory (the latter two of which are UB).
-    /// This isn't forbidden, however, because some tables rely on `T` being larger than a provided SDT in some
-    /// versions of ACPI (the [`ExtendedField`](crate::sdt::ExtendedField) type will be useful if you need to do
-    /// this. See our [`Fadt`](crate::fadt::Fadt) type for an example of this).
-    pub unsafe fn get_sdt<T>(&self, signature: sdt::Signature) -> Result<Option<PhysicalMapping<H, T>>, AcpiError>
-    where
-        T: AcpiTable,
-    {
-        let sdt = match self.sdts.get(&signature) {
-            Some(sdt) => sdt,
-            None => return Ok(None),
-        };
-        let mapping = unsafe { self.handler.map_physical_region::<T>(sdt.physical_address, sdt.length as usize) };
-
-        if !sdt.validated {
-            mapping.header().validate(signature)?;
+    /// Iterates through all of the SSDT tables.
+    pub fn iter_ssdts(&self) -> SsdtIterator<H> {
```

Comment:
Nit: I don't feel we need `iter` in the method name if it returns a `SsdtIterator` - `ssdts` feels nicer

Reply:
I feel `iter_ssdts` is more semantically valid, but I'm not super attached to the notion. Resolved!